File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> duplicates = new HashSet<>(); 
        int start = 0;
        int subLength = 0;

        for(int i=0; i< s.length(); i++){
            char current = s.charAt(i);

            while (duplicates.contains(current)) {
                duplicates.remove(s.charAt(start));
                start++;
            }
            duplicates.add(current);
            subLength = Math.max(subLength, i - start + 1);
        }
        return subLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        max_count = 0
        seen = {} #where key=char, val=index


        for r in range(len(s)):
            if s[r] in seen:
                l = max(l,seen[s[r]] + 1)
                
            
            #if s[r] not in seen,
            seen[s[r]] = r
            max_count = max(max_count, r-l+1)
        return max_count
            
        
            
        

        